students/timm67/session05/mailroom_part3.py

```diff
@@ -39,64 +39,63 @@ def send_thankyou_single_donor_task():
-        if donor[0] == donor_fname and donor[1] == donor_lname:
-            print("{0}, {1} FOUND".format(donor_fname, donor_lname))
-            donors[donor].append(donor_amt)
-            donor_found = True
-            break
-
-    if not donor_found:
-        print("{0}, {1} NOT found, creating".format(donor_fname, donor_lname))
-        donor = (donor_fname, donor_lname)
+    # Refactored and optimized to use a lot more efficient method
+    donor = (donor_fname, donor_lname)
+    try:
+        donors[donor].append(donor_amt)
+    except KeyError:
         donors[donor] = [donor_amt]
 
     fmtline0 = "\n\nDear {0},\n\n"
     fmtline1 = "Many thanks for your recent donation of {0:6.2f}.\n\n"
     fmtline2 = "Your total donations are {0:6.2f}, averaging {1:6.2f}\n\n"
     fmtline3 = "Thanks for your generous donation.\n\nBest Regards,\nThe Staff"
     donations = donors[donor]
-    total_donations = sum(donations)
+    # Can be done with list comprehension in one line, but reduction is needed
+    total_donations = reduce(add, donations)
     avg_donation = total_donations / len(donations)
     print(fmtline0.format(donor[0]))
     print(fmtline1.format(donations[(len(donations) - 1)]))
     print(fmtline2.format(total_donations, avg_donation))
     print(fmtline3.format())
 
 
-def send_thankyou_multiple_donors_task():
-    global donors
+def write_thankyou_to_file(donor):
     ffmtline0 = "\n\nDear {0},\n\n"
     ffmtline1 = "Many thanks for your recent donation of {0:6.2f}.\n\n"
     ffmtline2 = "Your total donations are {0:6.2f}, averaging {1:6.2f}\n\n"
     ffmtline3 = "Thanks for your generous donation.\nBest Regards,\nThe Staff"
-    for donor in donors.keys():
-        filename = "./{0}_{1}.txt".format(donor[0], donor[1])
-        print("Writing {0}...".format(filename))
-        donations = donors[donor]
-        total_donations = sum(donations)
-        avg_donation = total_donations / len(donations)
-        with open(filename, 'w') as handle:
-            handle.write(ffmtline0.format(donor[0]))
-            handle.write(ffmtline1.format(donations[(len(donations) - 1)]))
-            handle.write(ffmtline2.format(total_donations, avg_donation))
-            handle.write(ffmtline3.format())
+    filename = "./{0}_{1}.txt".format(donor[0], donor[1])
+    donations = donors[donor]
+    # Can be done with list comprehension in one line, but reduction is needed
+    total_donations = reduce(add, donations)
+    avg_donation = total_donations / len(donations)
+    with open(filename, 'w') as handle:
+        handle.write(ffmtline0.format(donor[0]))
+        handle.write(ffmtline1.format(donations[(len(donations) - 1)]))
+        handle.write(ffmtline2.format(total_donations, avg_donation))
+        handle.write(ffmtline3.format())
+
+
+def send_thankyou_multiple_donors_task():
+    global donors
+    # List comprehension used to write letters to multiple donors!
+    [write_thankyou_to_file(donor) for donor in donors.keys()]
```

Comment:
Not a good usage/pattern of list comprehension. Just a plain `for` loop is more desirable in cases like this.